pkg/server/router/merge/merge.go

```diff
@@ -0,0 +1,215 @@
+		}
+
+		data, err := json.Marshal(merged)
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshal merged object: %w", err)
+		}
+
+		return data, nil
+	}
+}
+
+var APIGropuListMerger = typedMergeFunc(func(lhs metav1.APIGroupList, rhs metav1.APIGroupList) (metav1.APIGroupList, error) {
+	lhs.Groups = append(lhs.Groups, slices.DeleteFunc(rhs.Groups, func(target metav1.APIGroup) bool {
+		return slices.ContainsFunc(lhs.Groups, func(g metav1.APIGroup) bool {
+			return g.Name == target.Name
+		})
+	})...)
+
+	return lhs, nil
+})
+
+var OpenAPIV3Merger = typedMergeFunc(func(lhs handler3.OpenAPIV3Discovery, rhs handler3.OpenAPIV3Discovery) (handler3.OpenAPIV3Discovery, error) {
+	maps.Copy(lhs.Paths, rhs.Paths)
+	return lhs, nil
+})
+
+type merger struct {
+	primary   http.Handler
+	secondary http.Handler
+
+	merger mergeFunc
+}
+
+// merge provides a handler which reconciles the responses between two handlers. If either handler returns with an
+// error state, only the response of the primary handler will be used (ie. if the primary handler succeeds and the
+// secondary fails only the primary response will be used). If either handler is nil the non-nil handler is returned,
+// or nil if both handlers are nil.
+func Merge(primary http.Handler, secondary http.Handler, f mergeFunc) http.Handler {
+	if primary != nil && secondary != nil {
+		return &merger{
+			primary:   primary,
+			secondary: secondary,
+			merger:    f,
+		}
+	}
+
+	if primary == nil {
+		return secondary
+	}
+
+	if secondary == nil {
+		return primary
+	}
+
+	return nil
+}
+
+func (m *merger) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	primaryRecorder := newRecorder()
+	m.primary.ServeHTTP(primaryRecorder, r)
+
+	secondaryRecorder := newRecorder()
+	m.secondary.ServeHTTP(secondaryRecorder, r)
+
+	w.Header().Set(headerContentType, contentTypeApplicationJSON)
+
+	if primaryRecorder.isOk() && !secondaryRecorder.isOk() {
+		b, err := io.ReadAll(primaryRecorder.body)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		if _, err := w.Write(b); err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		h := w.Header()
+		for k, v := range primaryRecorder.Header() {
+			h.Set(k, v[0])
+		}
+
+		w.WriteHeader(primaryRecorder.code)
+
+		return
+	}
+
+	primaryData, err := primaryRecorder.getBodyData()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	secondaryData, err := secondaryRecorder.getBodyData()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	merged, err := m.merger(primaryData, secondaryData)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if _, err := w.Write(merged); err != nil {
+		http.Error(w, fmt.Sprintf("failed to write response: %s", err), http.StatusInternalServerError)
+		return
+	}
+
+	// todo: write status
```

Comment:
Is there another issue for these TODOs?

Reply:
Oops, had these cleaned up in another branch that I didn't bring over. Should be good now!